### ISSUE TITLE
compress pdf file

### ISSUE THREAD

**Issue by cbjcbj** [gkovacs/pdfocr#28]
Hi, my input file of pdfocr is ~9M and my output file is about 390M, and I try to used pdftk to compress but the compress rate is less than 0.1%. So I wonder it is possible to compress the pdf file. Thank you.
![pic](https://cloud.githubusercontent.com/assets/20900795/20584243/ccc3ed34-b22b-11e6-85a3-d3b6bdc88073.png)

**Comment by wilsotc** [comment]
This problem is being caused by pdftoppm. I worked around it by bypassing this utility.

**Comment by cbjcbj** [comment]
So is it possible to convert ppm to jpg or something else and make the pdf file smaller?

**Comment by wilsotc** [comment]
Yes. You can often skip the ppm format step though. PDF allows you to
encode an image using other image formats including jpeg. The poppler
utility pdfimages extracts the PDF encoded image in its native format,
resolution, and color depth. When the image format isn't supported by your
OCR software, you can fall back to conversion. When it is supported,
there's no loss of PDF compression efficiency.

**Comment by cbjcbj** [comment]
Thank you. You said by skipping the ppm step. Do you mean I should change some lines of the ruby code or  I should add some command line parameters instead of `pdfocr -i input.pdf -o output.pdf`?

**Comment by wilsotc** [comment]
You would need to change the ruby code to bypass the pdftoppm utility. The
pdfimages utility is the more ideal route, but the imagemagick convert
utility could also be better. The bottom line is that the pdftoppm utility
is nearly worthless for monochrome pdf scanned documents.

**Comment by cbjcbj** [comment]
OK, thank you. I don't know ruby and I will have a try.

**Comment by wilsotc** [comment]
I have attached a modification of the pdfocr.rb that uses convert instead
of pdftoppm

**Comment by wilsotc** [comment]
`36d35
<   puts("running: " + c)
343c342
<   puts "Converting page #{i} to png"
---
>   puts "Converting page #{i} to ppm"
345,350c344,347
<   #sh "pdftoppm -r 300 #{shell_escape(basefn)}.pdf >#{shell_escape(basefn)}.ppm"
<   sh "convert -density 360 #{shell_escape(basefn)}.pdf -quality 35 #{shell_escape(basefn)}.png"
<   puts "convert -density 360 #{shell_escape(basefn)}.pdf -quality 35 #{shell_escape(basefn)}.png"
< 
<   if not File.file?(basefn+'.png')
<     puts "Error while converting page #{i} to png"
---
>   sh "pdftoppm -r 300 #{shell_escape(basefn)}.pdf >#{shell_escape(basefn)}.ppm"
>   puts "pdftoppm -r 300 #{shell_escape(basefn)}.pdf >#{shell_escape(basefn)}.ppm"
>   if not File.file?(basefn+'.ppm')
>     puts "Error while converting page #{i} to ppm"
356,357c353,354
<     sh "unpaper", basefn+'.png', basefn+'_unpaper.png'
<     if not File.file?(basefn+'_unpaper.png')
---
>     sh "unpaper", basefn+'.ppm', basefn+'_unpaper.ppm'
>     if not File.file?(basefn+'_unpaper.ppm')
361c358
<     sh "mv", basefn+'_unpaper.png', basefn+'.png'
---
>     sh "mv", basefn+'_unpaper.ppm', basefn+'.ppm'
366c363
<     sh "cuneiform", "-l", language, "-f", "hocr", "-o", basefn+'.hocr', basefn+'.png'
---
>     sh "cuneiform", "-l", language, "-f", "hocr", "-o", basefn+'.hocr', basefn+'.ppm'
368,370c365
<     #system("tesseract " + basefn+".png -l " + language + " -psm 0 " + basefn+"-new pdf")
<     #puts("tesseract " + basefn+".png -l " + language + " -psm 0 " + basefn+"-new pdf")
<     sh "tesseract", "-l", language, basefn+'.png', basefn+'-new', "pdf"
---
>     sh "tesseract", "-l", language, basefn+'.ppm', basefn+'-new', "pdf"
378c373
<     sh "rm -f #{tmp+'/'+'*.png'}"
---
>     sh "rm -f #{tmp+'/'+'*.ppm'}"
382c377
<     sh "ocroscript recognize #{shell_escape(basefn)}.png > #{shell_escape(basefn)}.hocr"
---
>     sh "ocroscript recognize #{shell_escape(basefn)}.ppm > #{shell_escape(basefn)}.hocr"
`

**Comment by wilsotc** [comment]
36d35
<   puts("running: " + c)
343c342
<   puts "Converting page #{i} to png"
---
>   puts "Converting page #{i} to ppm"
345,350c344,347
<   #sh "pdftoppm -r 300 #{shell_escape(basefn)}.pdf >#{shell_escape(basefn)}.ppm"
<   sh "convert -density 360 #{shell_escape(basefn)}.pdf -quality 35 #{shell_escape(basefn)}.png"
<   puts "convert -density 360 #{shell_escape(basefn)}.pdf -quality 35 #{shell_escape(basefn)}.png"
< 
<   if not File.file?(basefn+'.png')
<     puts "Error while converting page #{i} to png"
---
>   sh "pdftoppm -r 300 #{shell_escape(basefn)}.pdf >#{shell_escape(basefn)}.ppm"
>   puts "pdftoppm -r 300 #{shell_escape(basefn)}.pdf >#{shell_escape(basefn)}.ppm"
>   if not File.file?(basefn+'.ppm')
>     puts "Error while converting page #{i} to ppm"
356,357c353,354
<     sh "unpaper", basefn+'.png', basefn+'_unpaper.png'
<     if not File.file?(basefn+'_unpaper.png')
---
>     sh "unpaper", basefn+'.ppm', basefn+'_unpaper.ppm'
>     if not File.file?(basefn+'_unpaper.ppm')
361c358
<     sh "mv", basefn+'_unpaper.png', basefn+'.png'
---
>     sh "mv", basefn+'_unpaper.ppm', basefn+'.ppm'
366c363
<     sh "cuneiform", "-l", language, "-f", "hocr", "-o", basefn+'.hocr', basefn+'.png'
---
>     sh "cuneiform", "-l", language, "-f", "hocr", "-o", basefn+'.hocr', basefn+'.ppm'
368,370c365
<     #system("tesseract " + basefn+".png -l " + language + " -psm 0 " + basefn+"-new pdf")
<     #puts("tesseract " + basefn+".png -l " + language + " -psm 0 " + basefn+"-new pdf")
<     sh "tesseract", "-l", language, basefn+'.png', basefn+'-new', "pdf"
---
>     sh "tesseract", "-l", language, basefn+'.ppm', basefn+'-new', "pdf"
378c373
<     sh "rm -f #{tmp+'/'+'*.png'}"
---
>     sh "rm -f #{tmp+'/'+'*.ppm'}"
382c377
<     sh "ocroscript recognize #{shell_escape(basefn)}.png > #{shell_escape(basefn)}.hocr"
---
>     sh "ocroscript recognize #{shell_escape(basefn)}.ppm > #{shell_escape(basefn)}.hocr"

**Comment by wilsotc** [comment]
[fix.txt](https://github.com/gkovacs/pdfocr/files/684690/fix.txt)

**Comment by wilsotc** [comment]
This PERL script extracts all page images in their native format if they're JPG, PNG, or TIFF using the pdfimages utility. You could use it as the basis for a more capable converter.
[test.zip](https://github.com/gkovacs/pdfocr/files/684698/test.zip)

**Comment by cbjcbj** [comment]
Thank you, I will have a try:)

**Comment by wilsotc** [comment]
the syntax for the PERL script is:
test2.pl -i <pdffile>

**Comment by wodin** [comment]
A PDF I am trying this on actually has multiple images making up each page of the PDF.  I don't know what was used to scan the PDF, but most of these images are actually tiny PBMs corresponding to various small marks on the page.  The text is stored in one or a few PBMs or JPEGs.

For PDFs like this, I'm not sure whether it would be best to run the OCR engine on each image individually or to convert each page to a complete image as is done currently before running it through the OCR engine, but either way, I would prefer it if the text could be incorporated into a copy of the original PDF instead of using the exported images and the text to create the output.

I don't know how feasible this would be, but if possible, that seems like it would be a good way to do it.
